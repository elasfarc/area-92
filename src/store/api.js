// ACTION TYPES
export const API_CALL_REQUEST = "/api/request";
export const API_CALL_SUCCESS = "/api/success";
export const API_CALL_FAIL = "/api/fail";

// ACTION CREATOR
export const requestApiCall = (payload) => ({
  type: API_CALL_REQUEST,
  payload,
});
export const onApiSuccess = () => ({ type: API_CALL_SUCCESS });
export const onApiFail = () => ({ type: API_CALL_FAIL });
