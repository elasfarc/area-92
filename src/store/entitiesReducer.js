import { combineReducers } from 'redux'
import countryReducer from './country'

export default  combineReducers({
  countries: countryReducer
})