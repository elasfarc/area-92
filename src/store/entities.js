import * as apiActions from "./api";

// ACTION TYPES
const COUNTRIES_REQUESTED = "/countries/request";
const COUNTRIES_LOADED = "/countries/load";
const STATES_LOADED = "/country/states";
const CITIES_TRANSFORM_TO_GEO = "country/cities/transform2Geo";
const CITY_WEATHER_LOADED = "/weather/city";

// async thunk
export const loadCountriesData = () => async (dispatch) => {
  dispatch({ type: COUNTRIES_REQUESTED });
  try {
    const { data: countries } = await (
      await fetch(
        `https://countriesnow.space/api/v0.1/countries/info?returns=flag,currency,cities`
      )
    ).json();
    const { data: capitals } = await (
      await fetch(`https://countriesnow.space/api/v0.1/countries/capital`)
    ).json();
    dispatch(apiActions.onApiSuccess());
    dispatch({
      type: COUNTRIES_LOADED,
      payload: { countries, capitals },
    });
  } catch (error) {
    dispatch(apiActions.onApiFail());
  }
};

// ACTION CREATOR

export const getCountryStates = (country) =>
  apiActions.requestApiCall({
    url: `https://countriesnow.space/api/v0.1/countries/states`,
    method: "POST",
    body: { country },
    onStart: COUNTRIES_REQUESTED,
    onSuccess: STATES_LOADED,
    info: country,
  });

export const transformCitiesToGeo = ({ country, cities }) =>
  apiActions.requestApiCall({
    url: `http://open.mapquestapi.com/geocoding/v1/batch?key=4qiFlGPdSaIgG97k8C3AyJRKLBpq1jtJ`,
    method: "POST",
    body: { locations: cities.map((city) => ({ city, country })) },
    onStart: COUNTRIES_REQUESTED,
    onSuccess: CITIES_TRANSFORM_TO_GEO,
    info: country,
  });

export const getGeoWeather = () => ({});

// REDUCER
const initialState = {
  countries: [],
  isLoading: false,
};

const reducer = (state = initialState, { type, payload }) => {
  if (type === COUNTRIES_REQUESTED) return { ...state, isLoading: true };

  if (type === COUNTRIES_LOADED) {
    const { countries, capitals } = payload;
    return {
      ...state,
      isLoading: false,
      countries: countries.map((country) => ({
        ...country,
        ...capitals.find(({ capital, name }) => country.name === name),
      })),
    };
  }

  if (type === STATES_LOADED) {
    const {
      response: {
        data: { states },
      },
      info: countryName,
    } = payload;

    return {
      ...state,
      isLoading: false,
      countries: state.countries.map((country) =>
        country.name === countryName ? { ...country, states } : country
      ),
    };
  }

  if (type === CITIES_TRANSFORM_TO_GEO) {
    return {
      ...state,
      isLoading: false,
      countries: state.countries.map((country) =>
        country.name === payload.info
          ? {
              ...country,
              cities: payload.response.results.map(
                ({
                  providedLocation,
                  locations: [{ latLng, mapUrl, adminArea3 }],
                }) => ({
                  name: providedLocation.city,
                  country: country.name,
                  state: adminArea3,
                  latLng,
                  mapUrl,
                })
              ),
            }
          : country
      ),
    };
  }
  if (type === CITY_WEATHER_LOADED) return state;
  return state;
};

export default reducer;
