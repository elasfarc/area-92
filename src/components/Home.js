import React from "react";
import { Link, useRouteMatch } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import * as entitiesActions from "../store/entities";

const Home = () => {
  // const { url, path } = useRouteMatch();
  // console.log("url", url, "path", path, "+++", `${url}/africa`);
  const dispatch = useDispatch();
  const { continents } = useSelector((state) => state.entities);
  return (
    <div className="home">
      {Object.keys(continents).map((continent) => (
        <div key={continent} className="box mb-1">
          <Link
            to={`/${continent}`}
            onClick={() =>
              dispatch(
                entitiesActions.loadCountriesPerContinent(continent.toString())
              )
            }
          >
            <div className="container flex gap-2 cross-center">
              <img
                src={`${process.env.PUBLIC_URL}/imgs/${continent}.png`}
                alt=""
                className="box__img"
              />
              <div className="text-center">
                <h4 className="box__heading">{continent}</h4>
                <p className="box__text">blah blah blah</p>
              </div>
            </div>
          </Link>
        </div>
      ))}
    </div>
  );
};

export default Home;
