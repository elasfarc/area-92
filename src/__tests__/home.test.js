import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import Home from "../components/Home";
import store from "../store/configureStore";

const MockHome = () => (
  <Provider store={store}>
    <Home />
  </Provider>
);

describe("Mission component", () => {
  it("should match the snapshot", () => {
    const tree = renderer.create(<MockHome />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
