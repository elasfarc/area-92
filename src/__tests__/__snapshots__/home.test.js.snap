// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home(countries) component should match the snapshot 1`] = `
<h1
  className="text-center"
>
  Loading Countries...
</h1>
`;
